lib/azure/armrest/model/hash_model.rb

```diff
@@ -0,0 +1,140 @@
+require 'active_support/core_ext/string/inflections'
+
+module Azure
+  module Armrest
+    class HashModel
+      class << self
+        private
+
+        # Convert a key to a model class attribute which must be a qualified ruby method name
+        # key is from the original JSON or Hash object
+        #
+        # This is the default implementation, to be overridden by subclasses.
+        #
+        def key_to_attribute(key)
+          key.to_s.underscore
+        end
+
+        # Should setter methods be automatically created?
+        #
+        # Default to true, to be overridden by subclasses.
+        #
+        def generate_writer?
+          true
+        end
+
+        # Should values of this attribute be excluded from converting to model objects
+        #
+        # Default to false, to be overridden by subclasses.
+        #
+        def attr_excluded?(_attr_name)
+          false
+        end
+
+        # Declare attributes that need to be converted to Model
+        def attr_model(*attrs)
+          # Merge the declared attributes to the modelized list
+          @modelized_list = modelized_list | Set.new(attrs.map(&:to_s))
```

Comment:
Hmmm... since this is a class method, is this thread safe?

Reply:
We do the similar stuff in the current code, do we not?

Reply:
Not so much, but I will try an explain with an example.

Say we were making a call that made use of the `Parallel` gem to make multiple concurrent requests.  In the process of doing that, each of those are coming back at different times, instantiating there classes individually, and mutating `@modelized_list` at different points (with potentially different data).  This was not an issue with the previous implementation since any of the things being changed were on the instance level, and not shared memory of the class definition/class variables between multiple threads.

That said, with the `|` being done here, that might mitigate or even completely avoid issues because that would remove any duplicates being added.  This hurts my brain even to think about, let alone explain, so I could be over thinking this as well (part of the reason I brought this up as a question, not a statement 😉 )